rust/src/ldap/detect.rs

```diff
@@ -560,6 +564,293 @@ unsafe extern "C" fn ldap_tx_get_responses_msg(
+        return -1;
+    }
+    return 0;
+}
+
+unsafe extern "C" fn ldap_detect_request_attributes_get_data(
+    de: *mut c_void, transforms: *const c_void, flow: *const c_void, flow_flags: u8,
+    tx: *const c_void, list_id: c_int, local_id: u32,
+) -> *mut c_void {
+    return DetectHelperGetMultiData(
+        de,
+        transforms,
+        flow,
+        flow_flags,
+        tx,
+        list_id,
+        local_id,
+        ldap_tx_get_req_attributes,
+    );
+}
+
+unsafe extern "C" fn ldap_tx_get_req_attributes(
+    tx: *const c_void, _flags: u8, local_id: u32, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> bool {
+    let tx = cast_pointer!(tx, LdapTransaction);
+
+    *buffer = std::ptr::null();
+    *buffer_len = 0;
+
+    if let Some(request) = &tx.request {
+        let mut pos = 0_u32; // Contador para mapear local_id
+
+        match &request.protocol_op {
+            ProtocolOp::ModifyRequest(req) => {
+                for change in &req.changes {
+                    for val in &change.modification.attr_vals {
+                        if pos == local_id {
+                            let mut v = Vec::new();
+                            v.extend_from_slice(change.modification.attr_type.0.as_bytes());
+                            v.push(b'=');
+                            v.extend_from_slice(&val.0);
+                            tx.req_attributes.push(v);
+                            let value = tx.req_attributes.last().unwrap();
+
+                            *buffer = value.as_ptr();
+                            *buffer_len = value.len() as u32;
+                            return true;
+                        }
+                        pos += 1;
+                    }
+                }
+            }
+            ProtocolOp::AddRequest(req) => {
+                for attr in &req.attributes {
+                    for val in &attr.attr_vals {
+                        if pos == local_id {
+                            let mut v = Vec::new();
+                            v.extend_from_slice(attr.attr_type.0.as_bytes());
+                            v.push(b'=');
+                            v.extend_from_slice(&val.0);
+                            tx.req_attributes.push(v);
+                            let value = tx.req_attributes.last().unwrap();
+
+                            *buffer = value.as_ptr();
+                            *buffer_len = value.len() as u32;
+                            return true;
+                        }
+                        pos += 1;
+                    }
+                }
+            }
+            ProtocolOp::CompareRequest(req) => {
+                if local_id == 0 {
+                    let mut v = Vec::new();
+                    v.extend_from_slice(req.ava.attribute_desc.0.as_bytes());
+                    v.push(b'=');
+                    v.extend_from_slice(&req.ava.assertion_value);
+                    tx.req_attributes.push(v);
```

Comment:
we're not supposed to modify a tx in a detect keyword, why is this done?

Reply:
To save the buffer AttributeType=AttributeValue
@catenacyber was helping me with that, but I don't know if I'm on the right path

Reply:
Caching the value and storing it in the transaction

HTTP2 does that also

I feel it cleaner than `DetectRegisterThreadCtxGlobalFuncs`

Reply:
I missed that. I disagree with that pattern.

Reply:
Parser creates, updates tx, detect keywords consider it immutable.

In a case like this, if they keyword needs data in a special format, it should either be done in the parser, or it should be done in GetData in the inspect buffer

Reply:
> In a case like this, if they keyword needs data in a special format, 

yes

> it should either be done in the parser, 

But so, we waste resources for the people not using this keyword. (is it a problem ?)

> or it should be done in GetData in the inspect buffer

Well, that is what's being done, plus `GetData` stores the result in the transaction so that we do not have to compute it again, and it makes sense to me that it belongs there.

Do you mean `GetData` should compute the value and move its ownership to the InspectionBuffer ? And recompute it again on next packet (for a tx living multiple packets) ?